test/automation/src/profiler.ts

```diff
@@ -0,0 +1,267 @@
+		returnByValue: false,
+		objectGroup,
+	});
+	const objects = await driver.queryObjects({
+		prototypeObjectId: prototypeDescriptor.result.objectId!,
+		objectGroup,
+	});
+	const fnResult1 = await driver.callFunctionOn({
+		functionDeclaration: `function(){
+	const objects = this
+
+	const nativeConstructors = [
+		Object,
+		Array,
+		Function,
+		Set,
+		Map,
+		WeakMap,
+		WeakSet,
+		RegExp,
+		Node,
+		HTMLScriptElement,
+		DOMRectReadOnly,
+		DOMRect,
+		HTMLHtmlElement,
+		Node,
+		DOMTokenList,
+		HTMLUListElement,
+		HTMLStyleElement,
+		HTMLDivElement,
+		HTMLCollection,
+		FocusEvent,
+		Promise,
+		HTMLLinkElement,
+		HTMLLIElement,
+		HTMLAnchorElement,
+		HTMLSpanElement,
+		ArrayBuffer,
+		Uint16Array,
+		HTMLLabelElement,
+		TrustedTypePolicy,
+		Uint8Array,
+		Uint32Array,
+		HTMLHeadingElement,
+		MediaQueryList,
+		HTMLDocument,
+		TextDecoder,
+		TextEncoder,
+		HTMLInputElement,
+		HTMLCanvasElement,
+		HTMLIFrameElement,
+		Int32Array,
+		CSSStyleDeclaration
+	]
+
+	const isNativeConstructor = object => {
+		return nativeConstructors.includes(object.constructor) ||
+			object.constructor.name === 'AsyncFunction' ||
+			object.constructor.name === 'GeneratorFunction' ||
+			object.constructor.name === 'AsyncGeneratorFunction'
+	}
+
+	const isInstance = (object) => {
+		return object && !isNativeConstructor(object)
+	}
+
+	const instances = objects.filter(isInstance)
+	return instances
+}`,
+		objectId: objects.objects.objectId,
+		returnByValue: false,
+		objectGroup,
+	});
+
+	const fnResult2 = await getInstanceCountMap(driver, objectGroup, fnResult1.result);
+	const fnResult3 = await getInstanceCountArray(driver, objectGroup, fnResult2.result);
+	return {
+		value: fnResult3.result.value,
+		dispose: async () => {
+			// release object group
+			await driver.releaseObjectGroup({ objectGroup: objectGroup });
+		}
+	};
+};
+
+const getInstanceCountMap = async (driver: PlaywrightDriver, objectGroup: string, objects: Protocol.Runtime.RemoteObject) => {
+	const fnResult1 = await driver.callFunctionOn({
+		functionDeclaration: `function(){
+	const instances = this
+
+	const map = new Map()
+
+	for(const instance of instances){
+		if(map.has(instance.constructor)){
+			map.set(instance.constructor, map.get(instance.constructor) + 1)
+		} else {
+			map.set(instance.constructor, 1)
+		}
+	}
+	return map
+}`,
+		objectId: objects.objectId,
+		returnByValue: false,
+		objectGroup,
+	});
+
+	return fnResult1;
+};
+
+const getInstanceCountArray = async (driver: PlaywrightDriver, objectGroup: string, map: Protocol.Runtime.RemoteObject) => {
+	const fnResult1 = await driver.callFunctionOn({
+		functionDeclaration: `function(){
+	const map = this
+	const array = []
+
+	for(const [instanceConstructor, count] of map.entries()){
+		array.push({
+			name: instanceConstructor.name,
+			count,
+		})
+	}
+
+	return array
+}`,
+		objectId: map.objectId,
+		returnByValue: true,
+		objectGroup,
+	});
+	return fnResult1;
+};
```

Comment:
One could also implement the function to pass an array of classNames. It seems it could possibly simplify the code and improve performance:

```ts
const getInstances = async (driver: PlaywrightDriver, classNames:string[]): Promise<Array<{ name: string; count: number }>> => {
	await driver.collectGarbage();
	const objectGroup = `og:${generateUuid()}`;
	const prototypeDescriptor = await driver.evaluate({
		expression: 'Object.prototype',
		returnByValue: false,
		objectGroup,
	});
	const objects = await driver.queryObjects({
		prototypeObjectId: prototypeDescriptor.result.objectId!,
		objectGroup,
	});
	const fnResult1 = await driver.callFunctionOn({
		functionDeclaration: `function(){
	const objects = this
	const classNames = JSON.parse(${JSON.stringify(classNames)})

	const nativeConstructors = [
		Object,
		Array,
		Function,
		Set,
		Map,
		WeakMap,
		WeakSet,
		RegExp,
		Node,
		HTMLScriptElement,
		DOMRectReadOnly,
		DOMRect,
		HTMLHtmlElement,
		Node,
		DOMTokenList,
		HTMLUListElement,
		HTMLStyleElement,
		HTMLDivElement,
		HTMLCollection,
		FocusEvent,
		Promise,
		HTMLLinkElement,
		HTMLLIElement,
		HTMLAnchorElement,
		HTMLSpanElement,
		ArrayBuffer,
		Uint16Array,
		HTMLLabelElement,
		TrustedTypePolicy,
		Uint8Array,
		Uint32Array,
		HTMLHeadingElement,
		MediaQueryList,
		HTMLDocument,
		TextDecoder,
		TextEncoder,
		HTMLInputElement,
		HTMLCanvasElement,
		HTMLIFrameElement,
		Int32Array,
		CSSStyleDeclaration
	]

	const isNativeConstructor = object => {
		return nativeConstructors.includes(object.constructor) ||
			object.constructor.name === 'AsyncFunction' ||
			object.constructor.name === 'GeneratorFunction' ||
			object.constructor.name === 'AsyncGeneratorFunction'
	}

	const isInstance = (object) => {
		return object && !isNativeConstructor(object)
	}

	const instances = objects.filter(isInstance)

	const counts = Object.create(null)
	for(const instance of instances){
	  const name=instance.constructor.name
		if(classNames.includes(name)){
			counts[name]||= 0
			counts[name]++
		}
	}
	return counts
}`,
		objectId: objects.objects.objectId,
		returnByValue: true ,
		objectGroup,
	});

	return fnResult1.result.value;
};
```

Now the other functions `getInstanceCountArray` and  `getInstanceCountMap` are not needed anymore and the comparison in `checkLeaks` could be simplified as well.

The reason of this code was to count different classes with the same name differently, e.g. if there are two classes named `Abc` the counts for both classes wouldn't get mixed up. 

But looking at the code, it seems the way class counts are used here there probably aren't classes with the same name are measured (e.g. there is only one class `NotebookCellTextModel`).

Reply:
Makes good sense to me, I adopted the proposal in https://github.com/microsoft/vscode/pull/224547/commits/e4702fa7ced3149396b0953a8273f61a56fe2b74